### PR TITLE
[SPARK-26656][SQL] Benchmarks for date and timestamp functions

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

Added the following benchmarks:
- Extract components from timestamp like year, month, day and etc.
- Current date and time
- Date arithmetic like date_add, date_sub
- Format dates and timestamps
- Convert timestamps from/to UTC

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DateTimeBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ *   3. generate result:
+ *      SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain <this class>"
+ *      Results will be written to "benchmarks/DateTimeBenchmark-results.txt".
+ * }}}
+ */
+object DateTimeBenchmark extends SqlBasedBenchmark {
+  def doBenchmark(cardinality: Int, expr: String): Unit = {
+    spark.range(cardinality)
+      .selectExpr(expr)
+      .write.format("noop").save()
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DateTimeBenchmark.scala]: Single line is enough and shorter?
```scala
spark.range(cardinality).selectExpr(expr).write.format("noop").save()
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DateTimeBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ *   1. without sbt:
+ *      bin/spark-submit --class <this class> --jars <spark core test jar> <sql core test jar>
+ *   2. build/sbt "sql/test:runMain <this class>"
+ *   3. generate result:
+ *      SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain <this class>"
+ *      Results will be written to "benchmarks/DateTimeBenchmark-results.txt".
+ * }}}
+ */
+object DateTimeBenchmark extends SqlBasedBenchmark {
+  def doBenchmark(cardinality: Int, expr: String): Unit = {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DateTimeBenchmark.scala]: I know this is not a strict rule in benchmark suite. But, if you don't mind, `private` would be better for `doBenchmark` and `run`.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @MaxGekk . This is a useful benchmark. I love this.
So, can we have a more complete coverage? For example, `datediff` or `months_between` are not covered here. Please review the full list and add here, too.

cc @gatorsmile

File: sql/core/benchmarks/DateTimeBenchmark-results.txt [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+------------------------------------------------------------------------------------------------
+months_between wholestage off                 1653 / 1654          6.1         165.3       1.0X
+months_between wholestage on                  1630 / 1636          6.1         163.0       1.0X
+
+Java HotSpot(TM) 64-Bit Server VM 1.8.0_202-ea-b03 on Mac OS X 10.14.2
+Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
+window:                                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------
+window wholestage off                         1636 / 1661          0.6        1635.9       1.0X
+window wholestage on                        19997 / 20240          0.1       19997.4       0.1X
```
Comment (MaxGekk) [inline sql/core/benchmarks/DateTimeBenchmark-results.txt]: This is interesting. Performance drop ~x12. /cc @hvanhovell

File: sql/core/benchmarks/DateTimeBenchmark-results.txt [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+------------------------------------------------------------------------------------------------
+date_trunc WEEK wholestage off                 794 /  797         12.6          79.4       1.0X
+date_trunc WEEK wholestage on                  754 /  761         13.3          75.4       1.1X
+
+Java HotSpot(TM) 64-Bit Server VM 1.8.0_202-ea-b03 on Mac OS X 10.14.2
+Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
+date_trunc QUARTER:                      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------
+date_trunc QUARTER wholestage off             5261 / 5271          1.9         526.1       1.0X
+date_trunc QUARTER wholestage on              5145 / 5151          1.9         514.5       1.0X
```
Comment (MaxGekk) [inline sql/core/benchmarks/DateTimeBenchmark-results.txt]: Truncation to `QUARTER` is comparable slow. Need to re-run the benchmark with new implementation: https://github.com/apache/spark/pull/23641/files#diff-da60f07e1826788aaeb07f295fae4b8aL747

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DateTimeBenchmark.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DateTimeBenchmark.scala]: changed

**Review by srowen — APPROVED** [review]

**Review by hvanhovell — APPROVED** [review]
LGTM - Merging to master.

File: sql/core/benchmarks/DateTimeBenchmark-results.txt (hunk 3 above)
Comment (hvanhovell) [reply sql/core/benchmarks/DateTimeBenchmark-results.txt]: It is, let's address this in a follow-up.

**Comment by dongjoon-hyun** [comment]
Hi, @MaxGekk .
Could you make a followup as we discussed here?
- https://github.com/apache/spark/pull/23661#issuecomment-457871144

File: sql/core/benchmarks/DateTimeBenchmark-results.txt (hunk 3 above)
Comment (gatorsmile) [reply sql/core/benchmarks/DateTimeBenchmark-results.txt]: Do we have a JIRA to track this?

**Pull request closed without merging**